### PR TITLE
DM-33303: Tests for pipelines where Task and Butler disagree on DatasetType

### PR DESCRIPTION
**Pull request by timj**
Depends on lsst/daf_butler#635 and lsst/pipe_base#228

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/ctrl/mpexec/preExecInit.py [hunk 1]
```diff
@@ -125,19 +124,55 @@ def initializeDatasetTypes(self, graph, registerDatasetTypes=False):
...
-            datasetTypes.initOutputs,
-            datasetTypes.intermediates,
-            datasetTypes.outputs,
+
+        for datasetTypes, is_input, is_output in (
+            (datasetTypes.initIntermediates, True, True),
+            (datasetTypes.initOutputs, False, True),
+            (datasetTypes.intermediates, True, True),
+            (datasetTypes.outputs, False, True),
         ):
```
Comment (andy-slac) [inline python/lsst/ctrl/mpexec/preExecInit.py:133]: `is_output` is always `True` here, maybe remove it from loop variables? Or did you also want to check pure inputs here?

File: python/lsst/ctrl/mpexec/preExecInit.py [hunk 2]
```diff
@@ -125,19 +124,55 @@ def initializeDatasetTypes(self, graph, registerDatasetTypes=False):
...
+            if is_input:
+                # This DatasetType must be compatible on get.
+                is_compatible = datasetType.is_compatible_with(expected)
+            if is_output:
+                # Has to be able to be converted to expect type on put.
+                put_is_compatible = expected.is_compatible_with(datasetType)
+                if is_input:
+                    is_compatible = put_is_compatible and is_compatible
+                else:
+                    is_compatible = put_is_compatible
```
Comment (andy-slac) [inline python/lsst/ctrl/mpexec/preExecInit.py:148]: This looks more complicated than it should be. Maybe set `is_compatible` to `True` initially? Or is there a case when both `is_input` and `is_output` are false?

File: tests/test_simple_pipeline_executor.py [hunk 3]
```diff
@@ -59,27 +81,150 @@ def test_from_task_class(self):
...
+            lsst.daf.butler.DatasetType(
+                "output",
+                dimensions=self.butler.registry.dimensions.empty,
+                storageClass="StructuredDataList",
+            )
+        )
+
+        with self.assertRaises(ValueError) as cm:
+            executor.run(register_dataset_types=True)
+        self.assertIn("inconsistent with registry definition", str(cm.exception))
```
Comment (andy-slac) [inline tests/test_simple_pipeline_executor.py:226]: Maybe `assertRaisesRegex` instead of separate test for exception message?

**Review by andy-slac — APPROVED** [review]
Looks good, couple of minor comments.

File: python/lsst/ctrl/mpexec/preExecInit.py (hunk 1 above)
Comment (timj) [reply python/lsst/ctrl/mpexec/preExecInit.py:133]: Incredibly, I had not spotted that the `is_output` flag was always true. I had assumed when writing `_register_dataset_types` that there would be cases where input was true but outputs was false but seemingly I'm not making use of that. I suppose I should simplify the code below.

File: tests/test_simple_pipeline_executor.py (hunk 3 above)
Comment (timj) [reply tests/test_simple_pipeline_executor.py:226]: Thanks. Had forgotten about that.

File: python/lsst/ctrl/mpexec/preExecInit.py (hunk 2 above)
Comment (timj) [reply python/lsst/ctrl/mpexec/preExecInit.py]: Cleaned it up.

**Pull request merged**